src/liballoc/sync.rs

```diff
@@ -1071,6 +1067,129 @@ impl<T> Weak<T> {
             ptr: NonNull::new(usize::MAX as *mut ArcInner<T>).expect("MAX is not 0"),
         }
     }
+
+    /// Returns a raw pointer to the object `T` pointed to by this `Weak<T>`.
+    ///
+    /// It is up to the caller to ensure that the object is still alive when accessing it through
+    /// the pointer.
+    ///
+    /// The pointer may be NULL or be dangling in case the object has already been destroyed.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(weak_into_raw)]
+    ///
+    /// use std::sync::{Arc, Weak};
+    /// use std::ptr;
+    ///
+    /// let strong = Arc::new(42);
+    /// let weak = Arc::downgrade(&strong);
+    /// // Both point to the same object
+    /// assert!(ptr::eq(&*strong, Weak::as_raw(&weak)));
+    /// // The strong here keeps it alive, so we can still access the object.
+    /// assert_eq!(42, unsafe { *Weak::as_raw(&weak) });
+    ///
+    /// drop(strong);
+    /// // But not any more. We can do Weak::as_raw(&weak), but accessing the pointer would lead to
+    /// // undefined behaviour.
+    /// // assert_eq!(42, unsafe { *Weak::as_raw(&weak) });
+    /// ```
+    #[unstable(feature = "weak_into_raw", issue = "60728")]
+    pub fn as_raw(this: &Self) -> *const T {
+        match this.inner() {
+            None => ptr::null(),
+            // FIXME: Is this UB? The allocation must still live (we have our weak count in there),
+            // but the data itself might have been dropped in place by now, so the reference would
+            // point to uninitialized memory. That would be invalid reference. But how to get
+            // pointer to that memory?
+            Some(inner) => &inner.data,
```

Comment:
@RalfJung has an RFC to cover this kind of issue: https://github.com/rust-lang/rfcs/pull/2582. I think just taking a reference and allowing it to coerce like this code does is the "right" way currently?

Reply:
I am afraid there is no right way currently. The RFC also does not provide a syntax that you can already write now that would be deemed right in the future; I tried that but there was explicit opposition claiming there was not enough evidence that this would help.

Reply:
Oh, also even in my original form you should write `&inner.data as *const _`; the coercion here might happen *after* the block ended and that would be too late.

Reply:
OK, I think I'll do the reverse trick of what is used to compute the start of the structure. It'll work as long as `T` is sized, which is currently the case.